README.md
# cis_camera [![Build Status](https://travis-ci.com/tork-a/cis_camera.svg?branch=master)](https://travis-ci.org/tork-a/cis_camera)

This package contains ROS driver nodes for CIS Corporation ToF Camera Sensor DCC-RGBD1 with USB 3.0 interface.

## System Configurations

- OS / ROS
    - Ubuntu 16.04 / ROS Kinetic
    - Ubuntu 18.04 / ROS Melodic
- USB 3.0 Port
- CIS ToF Camera Sensor DCC-RGBD1

## CIS ToF Camera Sensor DCC-RGBD1

- Camera system
    - Dimensions : H:50mm x W:55mm x D:35mm ( Protruding parts are not included )
    - Weight : 110 g
    - Frame rate : 30 fps
- RGB camera
    - 1/3” CMOS image sensor
    - Global shutter operation
    - QVGA : 1280 x 960
    - M12 lens
- Depth / IR camera
    - 1/4” CCD image sensor
    - VGA : 640 x 480
    - M12 lens
    - Output : Depth and IR images
    - Depth sensing type: ToF ( Time-of-Flight )
    - Depth range
        - mode 0 : 300 - 5000 mm
        - mode 1 : 150 - 700 mm
- NIR light source
    - 2 Laser Diodes : 850nm / Class 1
- USB output
    - USB 3.0 micro B
    - UVC interface
    - Images
        - RGB : YUV422 - 1920 × 960
        - IR : Gray 16bit - 640 × 480
        - Depth : Gray 16bit - 640 × 480
- Power source
  - DC 12V, 3A

![CIS ToF Camera](doc/images/cis-tof-camera_dcc-rgbd1.jpg)

# Quick Start

## Installation

### Installing ROS

Install "ROS Desktop Full" on Ubuntu PC.

- ROS Kinetic for Ubuntu 16.04
    - http://wiki.ros.org/kinetic/Installation/Ubuntu
- ROS Melodic for Ubuntu 18.04
    - http://wiki.ros.org/melodic/Installation/Ubuntu

## Installation of cis_camera

```
$ source /opt/ros/$ROS_DISTRO/setup.bash
$ mkdir -p ~/camera_ws/src
$ cd ~/camera_ws/src
$ catkin_init_workspace
$ git clone https://github.com/tork-a/cis_camera.git
$ cd ~/camera_ws
$ rosdep install -y -r --from-paths src --ignore-src
$ catkin_make
$ source ~/camera_ws/devel/setup.bash
```

- **NOTE:** Replase `$ROS_DISTRO` to the ROS distribution of your system, `kinetic` or `melodic`.

## Device Permission Configuration

For the first time you start using CIS ToF camera, run `set_udev_rules` to set CIS ToF camera device permission configuration with entering sudo password in responce to program input.

```
$ source ~/camera_ws/devel/setup.bash
$ rosrun cis_camera set_udev_rules
```

- **NOTE** : This process is needed only once after the installations on your Ubuntu PC.

## Connecting Camera

1. Connect the camera to the USB 3.0 port of your Ubuntu PC
1. Connect the external power source to the camera and turn it on

- **NOTE** : It takes about 4 seconds for the camera to start up normally after the external power is turned on.

## Launching Software

### PointCloud

To see the pointcloud with RViz.

```
$ source ~/camera_ws/devel/setup.bash
$ roslaunch cis_camera pointcloud.launch
```

This file launches windows of RViz and `rqt_reconfigure`.

When you do not need to launch `rqt_reconfigure`, 
please set a launch option as below.

```
$ roslaunch cis_camera pointcloud.launch reconfigure:=false
```

![RViz PointCloud.launch](doc/images/cis_camera_pointcloud_rviz.png)

<img src="doc/images/cis_camera_dynamic_reconfigure.png" style="width: 50%;" />

#### Launch Options and Default Values of pointcloud.launch

- `rviz:=true`
    - Launching RViz 
- `reconfigure:=true`
    - Launching Dynamic Reconfigure
- `camera:=camera`
    - Name of cis_camera for ROS nodes and topics
- `num_worker_threads:=4`
    - Number of threads
- `vendor:=0x2af2`
    - Vendor ID of cis_camera
- `product:=0x1001`
    - Product ID of cis_camera
- `pointcloud_rgb:=false`
    - Projecting RGB colors on the pointcloud
- `flying_pixel_filter:=false`
    - Applying flying pixel filter with PCL `VoxelGrid` and `StatisticalOutlierRemoval` filters

![RGB PointCloud](doc/images/cis_camera_pointcloud_rgb.png)

### Publishing Images Only

When you publish only Depth, IR and RGB images, launch `tof.launch`.

```
$ source ~/camera_ws/devel/setup.bash
$ roslaunch cis_camera tof.launch
```

If you show the images, run `rqt` and open Plugins -> Visualization -> Image View.

```
$ source ~/camera_ws/devel/setup.bash
$ rqt
```

### Dynamic Reconfigure

After you launched `pointcloud.launch reconfigure:=false` or `tof.launch`, 
you can also reconfigure Depth/IR configurations dynamically with launching `rqt_reconfigure`.

```
$ source ~/camera_ws/devel/setup.bash
$ rosrun rqt_reconfigure rqt_reconfigure
```

When you reconfigure Depth/IR camera distortion correction parameters,
check `ir_dist_reconfig` to effect parameters `ir_fx`, `ir_fy` and so on.

To set back the parameters to `config/camera_ir.yaml` data, 
uncheck `ir_dist_reconfig`.

<img src="doc/images/cis_camera_rqt_reconfigure_check-ir_dist_reconfig.png" style="width: 50%;" />

### Frame Rate

When you want to know a frame rate of ROS topic, please run `rostopic hz` as below.

In the case of a topic `/camera/depth/points`,

```
$ source ~/camera_ws/devel/setup.bash
$ rostopic hz /camera/depth/points
```

To find out what topics exits,

```
$ source ~/camera_ws/devel/setup.bash
$ rostopic list
```

### Point Clud Library (PCL) Sample program

**Terminal 1**
```
$ source ~/camera_ws/devel/setup.bash
$ roslaunch cis_camera pointcloud.launch
```

**Terminal 2**
```
$ source ~/camera_ws/devel/setup.bash
$ rosrun cis_camera pcl_example
```

This PCL example code extracts a target object by filtering the point cloud, 
calculates the centroid of the extracted point cloud and publishes a TF on the centroid.

![PCL Example](doc/images/cis-camera_pcl-example_object-tf_clipped.png)

This example is based on "Building a Perception Pipleline" of ROS Industrial Training.

* https://industrial-training-master.readthedocs.io/en/melodic/_source/session5/Building-a-Perception-Pipeline.html
* https://industrial-training-master.readthedocs.io/en/kinetic/_source/session5/Building-a-Perception-Pipeline.html
* https://industrial-training-jp.readthedocs.io/ja/latest/_source/session5_JP/Building-a-Perception-Pipeline_JP.html


### Quit Software

Enter `Ctrl-C` on the running terminal.


<!-- EOF  -->


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package cis_camera
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.4 (2019-12-11)
------------------
* Merge pull request `#25 <https://github.com/tork-a/cis_camera/issues/25>`_ from tork-a/add-comments
  Add document and comments for document
* add 'Trouble Shooting' in index.rst
* fix style 2 by yy
* fix style
* Merge branch 'add-comments' of https://github.com/tork-a/cis_camera into add-comments
* remove doc/manifest.yaml
* Update index.rst
* mod index.rst format
* mod spec. list
* remove 'Fig.' for captions
* update documant index.rst, add build process for making document pdf file.
* update documents
* mod maintainer as dev@TORK and add author informaiton, add files for documents.
* remove doc/html/ folder
* doc/html generated by rosdoc_lite
* Merge branch 'master' into add-comments
* Merge pull request `#24 <https://github.com/tork-a/cis_camera/issues/24>`_ from tork-a/fix-zero-depth-cnv-gain
  Fix zero depth cnv gain and add RGB camera info dynamic_reconfigure
* add rgb camera info dynamic_reconfigure
* fix depth_cnv_gain\_ getting zero sometimes after the launch.
* Merge pull request `#23 <https://github.com/tork-a/cis_camera/issues/23>`_ from tork-a/add-frame-data-check
  mod num_worker_threads from 32 to 4, mod unexpected frame data size w…
* mod num_worker_threads from 32 to 4, mod unexpected frame data size warning message.
* Merge pull request `#22 <https://github.com/tork-a/cis_camera/issues/22>`_ from tork-a/add-frame-data-check
  Add frame data check, mod RGB pointcloud with tf
* add processing.launch.xml for RGB pointcloud with tf
* remove old code lines those are commented out
* add frame data size check to avoid memory segmentation error, add debug mode.
* add comments for documents
* Merge branch 'master' of https://github.com/tork-a/cis_camera into mod-cmakelists
* add tf_conversions in CMakeLists.txt
* Contributors: Tokyo Opensource Robotics Developer 534, Yosuke Yamamoto


src/main.cpp
// Copyright (c) 2019, Analog Devices Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
//
// * Redistributions of source code must retain the above copyright
//   notice, this list of conditions and the following disclaimer.
//
// * Redistributions in binary form must reproduce the above copyright
//   notice, this list of conditions and the following disclaimer in
//   the documentation and/or other materials provided with the
//   distribution.
//
// * Neither the name of the copyright holder nor the names of its
//   contributors may be used to endorse or promote products derived
//   from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
// COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
// INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
// OF THE POSSIBILITY OF SUCH DAMAGE.

// Based on original source from Ken Tossell's libuvc_camera. License copied below.
//
// * fix code style for ROS coding guidelines
// * change namespace from libuvc_camera to cis_camera

/*********************************************************************
* Software License Agreement (BSD License)
*
*  Copyright (C) 2012 Ken Tossell
*  All rights reserved.
*
*  Redistribution and use in source and binary forms, with or without
*  modification, are permitted provided that the following conditions
*  are met:
*
*   * Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*   * Redistributions in binary form must reproduce the above
*     copyright notice, this list of conditions and the following
*     disclaimer in the documentation and/or other materials provided
*     with the distribution.
*   * Neither the name of the author nor other contributors may be
*     used to endorse or promote products derived from this software
*     without specific prior written permission.
*
*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
*  POSSIBILITY OF SUCH DAMAGE.
*********************************************************************/


#include <ros/ros.h>
#include "cis_camera/camera_driver.h"


int main( int argc, char **argv )
{
  ros::init( argc, argv, "cis_camera" );
  ros::NodeHandle nh;
  ros::NodeHandle priv_nh( "~" );
  
  cis_camera::CameraDriver driver( nh, priv_nh );
  
  if ( !driver.Start() )
    return -1;
  
  ros::spin();
  
  driver.Stop();
  
  return 0;
}


src/nodelet.cpp
// Copyright (c) 2019, Analog Devices Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
//
// * Redistributions of source code must retain the above copyright
//   notice, this list of conditions and the following disclaimer.
//
// * Redistributions in binary form must reproduce the above copyright
//   notice, this list of conditions and the following disclaimer in
//   the documentation and/or other materials provided with the
//   distribution.
//
// * Neither the name of the copyright holder nor the names of its
//   contributors may be used to endorse or promote products derived
//   from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
// COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
// INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
// OF THE POSSIBILITY OF SUCH DAMAGE.

// Based on original source from Ken Tossell's libuvc_camera. License copied below.
//
// * fix code style for ROS coding guidelines
// * change namespace from libuvc_camera to cis_camera

/*********************************************************************
* Software License Agreement (BSD License)
*
*  Copyright (C) 2012 Ken Tossell
*  All rights reserved.
*
*  Redistribution and use in source and binary forms, with or without
*  modification, are permitted provided that the following conditions
*  are met:
*
*   * Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*   * Redistributions in binary form must reproduce the above
*     copyright notice, this list of conditions and the following
*     disclaimer in the documentation and/or other materials provided
*     with the distribution.
*   * Neither the name of the author nor other contributors may be
*     used to endorse or promote products derived from this software
*     without specific prior written permission.
*
*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
*  POSSIBILITY OF SUCH DAMAGE.
*********************************************************************/


#include <ros/ros.h>
#include <pluginlib/class_list_macros.h>
#include <nodelet/nodelet.h>

#include "cis_camera/camera_driver.h"

namespace cis_camera
{

class CameraNodelet : public nodelet::Nodelet
{
public:
  
  CameraNodelet() : running_(false) {}
  ~CameraNodelet();
  
private:
  
  virtual void onInit();
  
  volatile bool running_;
  boost::shared_ptr<CameraDriver> driver_;
};

CameraNodelet::~CameraNodelet()
{
  if (running_)
  {
    driver_->Stop();
  }
}

void CameraNodelet::onInit()
{
  ros::NodeHandle nh( getNodeHandle() );
  ros::NodeHandle priv_nh( getPrivateNodeHandle() );
  
  driver_.reset(new CameraDriver(nh, priv_nh));
  if (driver_->Start())
  {
    running_ = true;
  } 
  else
  {
    NODELET_ERROR("Unable to open camera.");
    driver_.reset();
  }
}

};

// Register this plugin with pluginlib.
//
// parameters are: class type, base class type
PLUGINLIB_EXPORT_CLASS( cis_camera::CameraNodelet, nodelet::Nodelet )


src/pcl_example.cpp
/*
 *  This example is based on "Building a Perception Pipleline" of ROS Industrial Training
 *  
 *  * https://industrial-training-master.readthedocs.io/en/melodic/_source/session5/Building-a-Perception-Pipeline.html
 *  * https://industrial-training-master.readthedocs.io/en/kinetic/_source/session5/Building-a-Perception-Pipeline.html
 *  * https://industrial-training-jp.readthedocs.io/ja/latest/_source/session5_JP/Building-a-Perception-Pipeline_JP.html
 *  
 */

#include <ros/ros.h>
#include <tf/transform_datatypes.h>
#include <tf/transform_listener.h>
#include <tf/transform_broadcaster.h>
#include <sensor_msgs/PointCloud2.h> //hydro

// PCL specific includes
#include <pcl_conversions/pcl_conversions.h> //hydro
#include "pcl_ros/transforms.h"

#include <pcl/filters/voxel_grid.h>
#include <pcl/filters/passthrough.h>
#include <pcl/sample_consensus/method_types.h>
#include <pcl/sample_consensus/model_types.h>
#include <pcl/segmentation/sac_segmentation.h>
#include <pcl/filters/extract_indices.h>
#include <pcl/segmentation/extract_clusters.h>
#include <pcl/filters/crop_box.h>
#include <pcl/filters/statistical_outlier_removal.h>
#include <tf_conversions/tf_eigen.h>
#include <pcl/segmentation/extract_polygonal_prism_data.h>


int main( int argc, char *argv[] )
{
  /*
   * INITIALIZE ROS NODE
   */
  ros::init( argc, argv, "pcl_example" );
  ros::NodeHandle nh;
  ros::NodeHandle priv_nh_( "~" );

  /*
   * SET UP PARAMETERS (COULD BE INPUT FROM LAUNCH FILE/TERMINAL)
   */
  std::string cloud_topic, world_frame, camera_frame;
  world_frame  = "map";
  camera_frame = "camera_depth";
  cloud_topic  = "/camera/depth/points";
  
  /*
   * SETUP PUBLISHERS
   */
  ros::Publisher object_pub, cluster_pub, pose_pub;
  object_pub  = nh.advertise<sensor_msgs::PointCloud2>( "object_cluster" , 1 );
  cluster_pub = nh.advertise<sensor_msgs::PointCloud2>( "primary_cluster", 1 );

  while (ros::ok())
  {
    /*
     * LISTEN FOR POINTCLOUD
     */
    std::string topic = nh.resolveName(cloud_topic);
    ROS_INFO_STREAM( "Cloud service called; waiting for a PointCloud2 on topic "<< topic );
    sensor_msgs::PointCloud2::ConstPtr recent_cloud =
                 ros::topic::waitForMessage<sensor_msgs::PointCloud2>(topic, nh);

    /*
     * TRANSFORM POINTCLOUD FROM CAMERA FRAME TO WORLD FRAME
     */
    tf::TransformListener listener;
    tf::StampedTransform stransform;
    try
    {
      listener.waitForTransform(world_frame, recent_cloud->header.frame_id,  ros::Time::now(), ros::Duration(6.0));
      listener.lookupTransform(world_frame, recent_cloud->header.frame_id,  ros::Time(0), stransform);
    }
    catch (tf::TransformException ex)
    {
      ROS_ERROR("%s",ex.what());
    }
    sensor_msgs::PointCloud2 transformed_cloud;
  //  sensor_msgs::PointCloud2::ConstPtr recent_cloud =
  //               ros::topic::waitForMessage<sensor_msgs::PointCloud2>(topic, nh);
    pcl_ros::transformPointCloud(world_frame, stransform, *recent_cloud, transformed_cloud);

    /*
     * CONVERT POINTCLOUD ROS->PCL
     */
    pcl::PointCloud<pcl::PointXYZ> cloud;
    pcl::fromROSMsg (transformed_cloud, cloud);
    
    /* ========================================
     * Fill Code: VOXEL GRID
     * ========================================*/
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_ptr (new pcl::PointCloud<pcl::PointXYZ> (cloud));
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_voxel_filtered (new pcl::PointCloud<pcl::PointXYZ> ());
    pcl::VoxelGrid<pcl::PointXYZ> voxel_filter;
    voxel_filter.setInputCloud (cloud_ptr);
    voxel_filter.setLeafSize ( float(0.01), float(0.01), float(0.01) );
    voxel_filter.filter (*cloud_voxel_filtered);
    
    //ROS_INFO_STREAM("Original cloud  had " << cloud_ptr->size() << " points");
    //ROS_INFO_STREAM("Downsampled cloud  with " << cloud_voxel_filtered->size() << " points");
    
    /* ========================================
     * Fill Code: PASSTHROUGH FILTER(S)
     * ========================================*/
    pcl::PointCloud<pcl::PointXYZ> xf_cloud, yf_cloud, zf_cloud;
    pcl::PassThrough<pcl::PointXYZ> pass_x;
    pass_x.setInputCloud(cloud_voxel_filtered);
    pass_x.setFilterFieldName("x");
    pass_x.setFilterLimits( -1.0, 0.5 );
    pass_x.filter(xf_cloud);

    pcl::PointCloud<pcl::PointXYZ>::Ptr xf_cloud_ptr(new pcl::PointCloud<pcl::PointXYZ>(xf_cloud));
    pcl::PassThrough<pcl::PointXYZ> pass_y;
    pass_y.setInputCloud(xf_cloud_ptr);
    pass_y.setFilterFieldName("y");
    pass_y.setFilterLimits( -0.3, 0.3 );
    pass_y.filter(yf_cloud);

    pcl::PointCloud<pcl::PointXYZ>::Ptr yf_cloud_ptr(new pcl::PointCloud<pcl::PointXYZ>(yf_cloud));
    pcl::PassThrough<pcl::PointXYZ> pass_z;
    pass_z.setInputCloud(yf_cloud_ptr);
    pass_z.setFilterFieldName("z");
    pass_z.setFilterLimits( -1.0, 3.0 );
    pass_z.filter(zf_cloud);
    
    /* ========================================
     * Fill Code: STATISTICAL OUTLIER REMOVAL (OPTIONAL)
     * ========================================*/
    pcl::PointCloud<pcl::PointXYZ>::Ptr zf_cloud_ptr( new pcl::PointCloud<pcl::PointXYZ>(zf_cloud));
    pcl::PointCloud<pcl::PointXYZ>::Ptr sor_cloud_filtered(new pcl::PointCloud<pcl::PointXYZ>);
    pcl::StatisticalOutlierRemoval<pcl::PointXYZ> sor;
    sor.setInputCloud( zf_cloud_ptr );
    sor.setMeanK ( 16 );
    sor.setStddevMulThresh ( 0.5 );
    sor.filter ( *sor_cloud_filtered );
    
    /* ========================================
     * Fill Code: CROPBOX (OPTIONAL)
     * ========================================*/
    
    /* ========================================
     * Fill Code: PLANE SEGEMENTATION
     * ========================================*/
    pcl::PointCloud<pcl::PointXYZ>::Ptr cropped_cloud(new pcl::PointCloud<pcl::PointXYZ>( *sor_cloud_filtered ));
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_f (new pcl::PointCloud<pcl::PointXYZ>);
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_filtered (new pcl::PointCloud<pcl::PointXYZ>);
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_plane (new pcl::PointCloud<pcl::PointXYZ> ());
    
    // Create the segmentation object for the planar model and set all the parameters
    pcl::SACSegmentation<pcl::PointXYZ> seg;
    pcl::PointIndices::Ptr inliers (new pcl::PointIndices);
    pcl::ModelCoefficients::Ptr coefficients (new pcl::ModelCoefficients);
    seg.setOptimizeCoefficients (true);
    seg.setModelType (pcl::SACMODEL_PLANE);
    seg.setMethodType (pcl::SAC_RANSAC);
    seg.setMaxIterations (100);
    seg.setDistanceThreshold (0.02);
    
    // Segment the largest planar component from the cropped cloud
    seg.setInputCloud (cropped_cloud);
    seg.segment (*inliers, *coefficients);
    if (inliers->indices.size () == 0)
    {
      ROS_WARN_STREAM ("Could not estimate a planar model for the given dataset.") ;
      //break;
    }
    
    // Extract the planar inliers from the input cloud
    pcl::ExtractIndices<pcl::PointXYZ> extract;
    extract.setInputCloud (cropped_cloud);
    extract.setIndices(inliers);
    extract.setNegative (false);
    
    // Get the points associated with the planar surface
    extract.filter (*cloud_plane);
    ROS_INFO_STREAM("PointCloud representing the planar component: " << cloud_plane->points.size () << " data points." );
    
    // Remove the planar inliers, extract the rest
    extract.setNegative (true);
    extract.filter ( *cloud_f );
    
    /* ========================================
     * Fill Code: PUBLISH PLANE MARKER (OPTIONAL)
     * ========================================*/
    
    
    /* ========================================
     * Fill Code: EUCLIDEAN CLUSTER EXTRACTION (OPTIONAL/RECOMMENDED)
     * ========================================*/
    
     
    // Creating the KdTree object for the search method of the extraction
    pcl::search::KdTree<pcl::PointXYZ>::Ptr tree (new pcl::search::KdTree<pcl::PointXYZ>);
    *cloud_filtered = *cloud_f;
    tree->setInputCloud (cloud_filtered);
    
    std::vector<pcl::PointIndices> cluster_indices;
    pcl::EuclideanClusterExtraction<pcl::PointXYZ> ec;
    ec.setClusterTolerance (0.01); // 2cm
    ec.setMinClusterSize (300);
    ec.setMaxClusterSize (10000);
    ec.setSearchMethod (tree);
    ec.setInputCloud (cloud_filtered);
    ec.extract (cluster_indices);

    std::vector<sensor_msgs::PointCloud2::Ptr> pc2_clusters;
    std::vector<pcl::PointCloud<pcl::PointXYZ>::Ptr > clusters;
    for (std::vector<pcl::PointIndices>::const_iterator it = cluster_indices.begin (); it != cluster_indices.end (); ++it)
    {
      pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_cluster (new pcl::PointCloud<pcl::PointXYZ>);
      for (std::vector<int>::const_iterator pit = it->indices.begin (); pit != it->indices.end (); pit++)
        cloud_cluster->points.push_back(cloud_filtered->points[*pit]);
      cloud_cluster->width = cloud_cluster->points.size ();
      cloud_cluster->height = 1;
      cloud_cluster->is_dense = true;
      std::cout << "Cluster has " << cloud_cluster->points.size() << " points.\n";
      clusters.push_back(cloud_cluster);
      sensor_msgs::PointCloud2::Ptr tempROSMsg(new sensor_msgs::PointCloud2);
      pcl::toROSMsg(*cloud_cluster, *tempROSMsg);
      pc2_clusters.push_back(tempROSMsg);
    }
    
    pcl::PointCloud<pcl::PointXYZ>::Ptr largest_cluster (new pcl::PointCloud<pcl::PointXYZ>);
    *largest_cluster = *(clusters.at(0));
    
    /* ========================================
     * Fill Code: PUBLISH OTHER MARKERS (OPTIONAL)
     * ========================================*/
    
    
    /* ========================================
     * BROADCAST TRANSFORM (OPTIONAL)
     * ========================================*/
    static tf::TransformBroadcaster br;
    tf::Transform part_transform;

    // Calculate Centroid Largest PointCloud Cluster
    Eigen::Vector4f xyz_centroid;
    pcl::compute3DCentroid( *largest_cluster, xyz_centroid );
    
    part_transform.setOrigin( tf::Vector3( xyz_centroid[0] , xyz_centroid[1], xyz_centroid[2] ) );
    tf::Quaternion q;
    q.setRPY(0, 0, 0);
    part_transform.setRotation(q);
    
    br.sendTransform(tf::StampedTransform(part_transform, ros::Time::now(), world_frame, "part"));
    
    /* ========================================
     * Fill Code: POLYGONAL SEGMENTATION (OPTIONAL)
     * ========================================*/
    
    
    /* ========================================
     * CONVERT POINTCLOUD PCL->ROS
     * PUBLISH CLOUD
     * Fill Code: UPDATE AS NECESSARY
     * ========================================*/
    sensor_msgs::PointCloud2::Ptr pc2_cloud (new sensor_msgs::PointCloud2);
    pcl::toROSMsg( *largest_cluster, *pc2_cloud );
    pc2_cloud->header.frame_id=world_frame;
    pc2_cloud->header.stamp=ros::Time::now();
    object_pub.publish(pc2_cloud);
  }
  
  return 0;
}

